src/audio/rtnr/rtnr.c

```diff
@@ -0,0 +1,546 @@
+	free_RtkMemoryPool();
+
+	rfree(cd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int rtnr_params(struct comp_dev *dev, struct sof_ipc_stream_params *params)
+{
+	int ret;
+
+	comp_info(dev, "rtnr_params()");
+
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "rtnr_params() error: comp_verify_params() failed.");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int rtnr_cmd_get_data(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata, int max_size)
+{
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "rtnr_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_get_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int rtnr_cmd_set_data(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata)
+{
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "rtnr_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_set_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int32_t rtnr_cmd_get_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		comp_dbg(dev, "rtnr_cmd_get_value(), SOF_CTRL_CMD_SWITCH, cdata->comp_id = %u",
+			 cdata->comp_id);
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_get_value() error: invalid cdata->cmd %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int32_t rtnr_cmd_set_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
+{
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		comp_dbg(dev, "rtnr_cmd_set_value(), SOF_CTRL_CMD_SWITCH, cdata->comp_id = %u",
+			 cdata->comp_id);
+		break;
+
+	default:
+		comp_err(dev, "rtnr_cmd_set_value() error: invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int rtnr_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "rtnr_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = rtnr_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = rtnr_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	case COMP_CMD_SET_VALUE:
+		ret = rtnr_cmd_set_value(dev, cdata);
+		break;
+	case COMP_CMD_GET_VALUE:
+		ret = rtnr_cmd_get_value(dev, cdata);
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd() error: invalid command");
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int rtnr_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "rtnr_trigger() cmd: %d", cmd);
+
+	return comp_set_state(dev, cmd);
+}
+
+/* copy and process stream data from source to sink buffers */
+static int rtnr_copy(struct comp_dev *dev)
```

Comment:
Agree to cujomalainey@ that there are too many burdens on the copy step (please consider that the copy step will be called per the consumption of audio buffers). Any mathematics should be pre-calculated which is not dependent on the copy state.

Reply:
The pre-calculated variables could be calculated on prepare() and stored in comp_data

Reply:
Ok the copy function now use single source stream now. However, the processing function uses multiple source streams as parameter. So we still need to put the source stream into an array before calling it.